tiatoolbox/models/engine/patch_predictor.py

```diff
@@ -721,49 +707,81 @@ def _predict_tile_wsi(
                 outputs.append(merged_prediction)
 
             if save_output:
-                # dynamic 0 padding
-                img_code = f"{idx:0{len(str(len(imgs)))}d}"
-
-                save_info = {}
-                save_path = os.path.join(str(save_dir), img_code)
-                raw_save_path = f"{save_path}.raw.json"
-                save_info["raw"] = raw_save_path
-                save_as_json(output_model, raw_save_path)
-                if merge_predictions:
-                    merged_file_path = f"{save_path}.merged.npy"
-                    np.save(merged_file_path, merged_prediction)
-                    save_info["merged"] = merged_file_path
-                file_dict[str(img_path)] = save_info
+                img_id, save_info = self._save_output(
+                    output_model,
+                    idx,
+                    merged_prediction,
+                    input_img,
+                    input_imgs,
+                    save_dir,
+                    merge_predictions,
+                )
+                file_dict[img_id] = save_info
 
         return file_dict if save_output else outputs
 
+    @staticmethod
+    def _save_output(
+        output_model,
+        idx,
+        merged_prediction,
+        input_img,
+        input_imgs,
+        save_dir,
+        merge_predictions,
+    ):
+        """Save prediction to json and/or numpy file."""
+        # dynamic 0 padding
+        img_code = f"{idx:0{len(str(len(input_imgs)))}d}"
+
+        save_info = {}
+        save_path = os.path.join(str(save_dir), img_code)
+        raw_save_path = f"{save_path}.raw.json"
+        save_info["raw"] = raw_save_path
+        save_as_json(output_model, raw_save_path)
+        if merge_predictions:
+            merged_file_path = f"{save_path}.merged.npy"
+            np.save(merged_file_path, merged_prediction)
+            save_info["merged"] = merged_file_path
+
+        img_id = None
+        if isinstance(input_img, WSIReader):
+            img_id = str(input_img.input_path)
+
+        elif isinstance(input_img, (str, pathlib.Path)):
+            img_id = str(input_img)
+
+        if img_id is None:
+            img_id = idx
+
+        return img_id, save_info
+
     def predict(
         self,
-        imgs,
```

Comment:
Please do not update these as these need to be consistent with SemanticSegmentor and other engines. This will be fixed in redesigned engines #578

Reply:
Now requirements.txt is identical to develop